### PR TITLE
[Yaml] Improve YAML boolean escaping

### PR DESCRIPTION
**Pull request by petert82**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #13209 |
| License | MIT |
| Doc PR | None |

This PR ensures that PHP [values which would be interpreted as booleans](http://yaml.org/type/bool.html) in older versions of the YAML spec are escaped with single quotes when dumped by the Dumper.

For example, dumping this:

``` php
array(
    'country_code' => 'no',
    'speaks_norwegian' => 'y',
    'heating' => 'on',
)
```

Will produce this YAML:

``` yaml
country_code: 'no'
speaks_norwegian: 'y'
heating: 'on'
```

### CODE REVIEW

File: src/Symfony/Component/Yaml/Escaper.php [hunk 1]
```diff
@@ -86,4 +86,28 @@ public static function escapeWithSingleQuotes($value)
...
+    /**
+     * Determines if a PHP value is entirely composed of a value that would
+     * require require single quoting in YAML.
+     *
+     * @param  string $value A PHP value
+     * @return bool          True if the value would require single quotes.
+     */
+    private static function isValueRequiresSingleQuoting($value)
+    {
+        return in_array(strtolower($value), array('null', '~', 'true', 'false', 'y', 'n', 'yes', 'no', 'on', 'off'));
```
Comment (fabpot) [inline src/Symfony/Component/Yaml/Escaper.php]: Note to myself and others: even if we do not support 'y', 'n', ... as valid Booleans, it does not hurt to escape them here for interoperability.

Comment (stof) [inline src/Symfony/Component/Yaml/Escaper.php]: I suggest adding it as a comment to avoid removing it by mistake in the future

**Comment by stof** [comment]
> One other issue we've seen between pecl/yaml and symfony/yaml is that - in a sequence is treated as a nested sequence by pecl/yaml but as a string by symfony/yaml

Can you open an issue for that ?

**Comment by larowlan** [comment]
> Can you open an issue for that ?
Let me confirm it first, I suspect it might have been hand-edited, but if not - will open

File: src/Symfony/Component/Yaml/Escaper.php [hunk 2]
```diff
@@ -86,4 +86,30 @@ public static function escapeWithSingleQuotes($value)
...
+    }
+
+    /**
+     * Determines if a PHP value is entirely composed of a value that would
+     * require single quoting in YAML.
+     *
+     * @param  string $value A PHP value
+     * @return bool          True if the value would require single quotes.
+     */
+    private static function isValueRequiresSingleQuoting($value)
```
Comment (fabpot) [inline src/Symfony/Component/Yaml/Escaper.php]: `doesValueRequireSingleQuoting`?

Comment (fabpot) [inline src/Symfony/Component/Yaml/Escaper.php]: Actually, I'm wondering if we really need those 2 private methods. What about inlining them with the right comments instead? Especially as they can be called quite a lot.

**Comment by fabpot** [comment]
:+1: after my small comment is addressed.

**Comment by petert82** [comment]
@fabpot They're gone. :)

**Comment by fabpot** [comment]
Thank you @petert82.

File: src/Symfony/Component/Yaml/Escaper.php [hunk 3]
```diff
@@ -72,7 +72,15 @@ public static function escapeWithDoubleQuotes($value)
...
-        return preg_match('/[ \s \' " \: \{ \} \[ \] , & \* \# \?] | \A[ \- ? | < > = ! % @ ` ]/x', $value);
+        // Determines if the PHP value contains any single characters that would
+        // cause it to require single quoting in YAML.
+        if (preg_match('/[ \s \' " \: \{ \} \[ \] , & \* \# \?] | \A[ \- ? | < > = ! % @ ` ]/x', $value)) {
+            return true;
+        }
+
+        // Determines if a PHP value is entirely composed of a value that would
+        // require single quoting in YAML.
+        return in_array(strtolower($value), array('null', '~', 'true', 'false', 'y', 'n', 'yes', 'no', 'on', 'off'));
```
Comment (stof) [inline src/Symfony/Component/Yaml/Escaper.php]: I would do the in_array check before the regex check. It is likely to be faster

File: src/Symfony/Component/Yaml/Inline.php [hunk 4]
```diff
@@ -135,12 +135,10 @@ public static function dump($value, $exceptionOnInvalidType = false, $objectSupp
             case Escaper::requiresDoubleQuoting($value):
                 return Escaper::escapeWithDoubleQuotes($value);
             case Escaper::requiresSingleQuoting($value):
+            case preg_match(self::getTimestampRegex(), $value):
                 return Escaper::escapeWithSingleQuotes($value);
             case '' == $value:
```
Comment (stof) [inline src/Symfony/Component/Yaml/Inline.php]: I suggest moving this fast comparison before the heavier regex checks

**Pull request merged**